gadget/validate.go

```diff
@@ -0,0 +1,72 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+func validateVolumeContentsPresence(gadgetSnapRootDir string, vol *LaidOutVolume) error {
+	// bare structure content is checked to exist during layout
+	// make sure that filesystem content source paths exist as well
+	for _, s := range vol.LaidOutStructure {
+		if s.IsBare() {
+			continue
+		}
+		for _, c := range s.Content {
+			realSource := filepath.Join(gadgetSnapRootDir, c.Source)
+			if !osutil.FileExists(realSource) {
+				return fmt.Errorf("cannot validate structure %v, content %v: source path does not exist", s, c)
```

Comment:
if I read the code correctly:  we'll get "cannot validate... cannot validate" error messages... in general we don't use "cannot validate" form of message, more like "invalid ..."